barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -509,4 +514,95 @@ namespace avm_alu(256);
+    // ========= SHIFT LEFT/RIGHT OPERATIONS ===============================
+    // Given inputs to a shift operation, a & b, and a memory tag, mem_tag.
+    // Split a into Big Endian hi and lo limbs, a_hi and a_lo, and the number of bits represented by the memory tag, t.
+    // QUESTION: SHOULD B BE CONSTRAINED TO BE U8 -> i.e. when would we shift more than 255 bits when the max number of bits of a is 128bits?
+    // If we are shifting by more than the bit length represented by the memory tag, the result is trivially zero
+    //
+    // === Steps when performing SHR
+    // (1) Prove the correct decomposition: a_hi * 2**b + a_lo = a
+    // (2) Range check a_hi < 2**(t-b) && a_lo < 2**b, ensure we have not overflowed the limbs during decomp
+    // (3) Return a_hi
+    //
+    //  <--(t-b) bits --> |   <-- b bits -->
+    // -------------------|-------------------
+    // |      a_hi        |       a_lo       | --> a
+    // ---------------------------------------
+    //
+    // === Steps when performing SHL
+    // (1) Prove the correct decomposition: a_hi * 2**(t-b) + a_lo = a
+    // (2) Range check a_hi < 2**b && a_lo < 2**(t-b)
+    // (3) Return a_lo * 2**b
+    // 
+    //  <-- b bits -->   | <-- (t-b) bits -->
+    // ------------------|-------------------
+    // |      a_hi       |      a_lo        | --> a
+    // --------------------------------------
+
+    // TODO: Possibly optimised with variable length checks
+    // Check that a_lo and a_hi are range checked such that that:
+    // SHR: a_hi < 2**(t - b) && a_lo < 2**b
+    // SHL: a_hi < 2**b && a_lo < 2**(t - b)
+
+    // In lieu of a variable length check, we can utilise 2 fixed range checks instead.
+    // Given the dynamic range check of 0 <= a_hi <= 2**(t-b), where b < t
+    // (1) 0 <= a_hi <= 2**t
+    // (2) 0 <= 2**(t - b)  - a_hi <= 2**t
+    // Note that (1) is guaranteed elsewhere through the tagged memory model, so we focus on (2) here. 
+
+    // === General Notes:
+    // There are probably ways to merge various relations for the SHL/SHR, but they are separate
+    // now while we are still figuring out.
+
+    // Indicate if the shift amount < MAX_BITS
+    pol commit shift_lt_bit_len;
+    shift_lt_bit_len * (1 - shift_lt_bit_len) = 0;
+
+    // The number of bits represented by the memory tag, any shifts greater than this will result in zero.
+    pol MAX_BITS = u8_tag * 8 + 
+                u16_tag * 16 +
+                u32_tag * 32 +
+                u64_tag * 64 +
+                u128_tag * 128;
+
+    // The result of MAX_BITS - ib
+    pol commit t_sub_b_bits;
+
+    // For our assumptions to hold, we must check that b < MAX_BITS. This can be achieved by the following relation.
+    // We check if b < MAX_BITS || b >= MAX_BITS using boolean shift_le_bit_len. 
+    pol SHIFT_DIFF = shift_lt_bit_len * (MAX_BITS - ib - 1) + (1 - shift_lt_bit_len) * (ib - MAX_BITS);
+
+    // Regardless of which side is evaluated, the value of t_sub_b_bits < 2**8
```

Comment:
We have to comment that ib is less than 2^128. Otherwise, if ib is very close to p we might get a wrong evaluation for shift_lt_bit_len due to an underflow of MAX_BITS - ib which would be so large to end up in valid lookup range, i.e., < 2**8

Reply:
Yep, I was saying that it is actually fine even when b is u128. It is then also the case if we restrict it to be u8. My suggestion is to add an explanation about this.